### PR TITLE
[query-frontend] Set write deadline for `/active_series` requests

### PR DESCRIPTION
**Pull request by flxbk**
#### What this PR does

The query-frontend uses the value supplied via `-server.http-write-timeout` (default `2m`) to close connections that have not finished writing the response by the deadline. The `/active_series` endpoint can generate large responses, the transmission of which can exceed that deadline. Therefore, this PR introduces a parameter to configure write timeouts for streaming responses (currently only applicable to the `/active_series` endpoint).

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated.
- [ ] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [x] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

File: pkg/frontend/transport/handler.go [hunk 1]
```diff
@@ -190,6 +190,17 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	activityIndex := f.at.Insert(func() string { return httpRequestActivity(r, r.Header.Get("User-Agent"), params) })
 	defer f.at.Delete(activityIndex)
 
+	if isStreamingEndpoint(r) {
+		// Disable the write deadline for streaming endpoints.
+		rc := http.NewResponseController(w)
+		err = rc.SetWriteDeadline(time.Time{})
```
Comment (Logiraptor) [inline pkg/frontend/transport/handler.go:196]: I'm not familiar with how this works, what happens if a client initiates a request and then just hangs without reading data? Does the connection eventually close or just sit around?

Comment (flxbk) [reply pkg/frontend/transport/handler.go:196]: Yes, if the deadline is set to the default value the connection will remain open until either
- the response has been transmitted in full
- the client closes it
- some hop in between (e.g. reverse proxy) closes it.

We could instead add a config parameter and use it to set a non-default deadline that's far enough in the future.

File: pkg/frontend/transport/handler.go [hunk 2]
```diff
@@ -53,13 +53,15 @@ type HandlerConfig struct {
 	LogQueryRequestHeaders flagext.StringSliceCSV `yaml:"log_query_request_headers" category:"advanced"`
 	MaxBodySize            int64                  `yaml:"max_body_size" category:"advanced"`
 	QueryStatsEnabled      bool                   `yaml:"query_stats_enabled" category:"advanced"`
+	ResponseStreamTimeout  time.Duration          `yaml:"response_stream_timeout" category:"experimental"`
```
Comment (pracucci) [inline pkg/frontend/transport/handler.go:56]: Have you considered to name the config parameter to be specific for active series endpoints? Reason is that if in the future we add streaming response to other endpoints, I would argue a 10m default timeout for them may be very high and undesiderable.

File: pkg/frontend/transport/handler.go [hunk 3]
```diff
@@ -190,6 +192,20 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	activityIndex := f.at.Insert(func() string { return httpRequestActivity(r, r.Header.Get("User-Agent"), params) })
 	defer f.at.Delete(activityIndex)
 
+	if isStreamingEndpoint(r) {
+		// Set write deadline for streaming responses.
+		var deadline time.Time
+		if f.cfg.ResponseStreamTimeout > 0 {
+			deadline = time.Now().Add(f.cfg.ResponseStreamTimeout)
```
Comment (pracucci) [inline pkg/frontend/transport/handler.go:199]: I wouldn't set a no deadline if `ResponseStreamTimeout` is 0. On the contrary, the whole `if isStreamingEndpoint(r)` should be skipped if `ResponseStreamTimeout == 0` and then we document that if it's zero the global timeout will be used. I think it's a bad practice allowing to have no timeout (even if someone could set a very high timeout, but at least it's more clear the intention).

File: pkg/frontend/transport/handler_test.go [hunk 4]
```diff
@@ -533,6 +533,47 @@ func TestHandler_LogsFormattedQueryDetails(t *testing.T) {
...
+	for _, tt := range []struct {
+		name      string
+		path      string
+		wantError bool
+	}{
+		{name: "deadline exceeded for non-streaming endpoint", path: "/api/v1/query", wantError: true},
+		{name: "deadline not exceeded for streaming endpoint", path: "/api/v1/cardinality/active_series"},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			const serverWriteTimeout = 10 * time.Millisecond
```
Comment (pracucci) [inline pkg/frontend/transport/handler_test.go:546]: I would increase it. Being very short, it could cause some flakyness on a slow CI.

File: pkg/util/gziphandler/gzip.go [hunk 5]
```diff
@@ -158,6 +158,10 @@ func (w *GzipResponseWriter) Write(b []byte) (int, error) {
 	return w.startPlainWrite(len(b))
 }
 
+func (w *GzipResponseWriter) Unwrap() http.ResponseWriter {
```
Comment (pracucci) [inline pkg/util/gziphandler/gzip.go:161]: Can you explain why is this change required?

File: pkg/frontend/transport/handler.go [hunk 6]
```diff
@@ -190,6 +192,20 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		}
+		err = http.NewResponseController(w).SetWriteDeadline(deadline)
```
Comment (pracucci) [inline pkg/frontend/transport/handler.go:201]: Can you confirm we don't need to also set the read deadline, right?

File: pkg/frontend/transport/handler.go (hunk 2 above)
Comment (pracucci) [reply pkg/frontend/transport/handler.go:56]: Given this looks applied only to write timeout, we could rename it to "active-series-write-timeout" to be more specific.

File: pkg/util/gziphandler/gzip.go (hunk 5 above)
Comment (flxbk) [reply pkg/util/gziphandler/gzip.go:163]: I've added a comment that explains why this is required.

File: pkg/frontend/transport/handler.go (hunk 6 above)
Comment (flxbk) [reply pkg/frontend/transport/handler.go:197]: Correct, the requests are typically small and the default read deadline should be applied.

File: pkg/frontend/transport/handler.go (hunk 3 above)
Comment (flxbk) [reply pkg/frontend/transport/handler.go]: I have changed the behaviour to apply the longer timeout only if it's configured and otherwise default to the same timeout as the other endpoints.

**Review by pracucci — APPROVED** [review]
LGTM! Thanks for addressing my feedback
